### ISSUE TITLE
Efficient compilation to loss backends

### ISSUE THREAD

**Issue by MatthewDaggitt** [vehicle-lang/vehicle#693]
Okay, so as we've found out our model of computation for vectors isn't a good match for Tensorflow which causes major efficiency problems when converting to loss functions.

As discussed, we need to make `Tensor` it's own builtin class. The best approach we think is to make tensors a wrapper around multi-dimensional vectors, i.e. have a constructor `tensor` that maps multi-dimensional vectors to tensors and have a constraint on the base element type.

To start with we'll make the user insert the `tensor` coercion manually. This can be relaxed later, if we figure out the necessary instance magic.

**Comment by MatthewDaggitt** [comment]
Okay, the above approach isn't feasible as `Tensor` infects absolutely everything.  The reason is that the lookup operation on tensors has to return a tensor with one less dimension. Therefore when looking up a value in a 1D tensor you get a 0D tensor back rather than something of the tensor element type.

I think we're going to have to perform another type-system based analysis pass to automatically translate to tensors... Going to see with what I can come up with...

**Comment by MatthewDaggitt** [comment]
Fixed by #832
